### ISSUE TITLE
Serializing complex generic class into json throws incompatibility error on deserialization

### ISSUE THREAD

**Issue by schmidt-florian** [FlaxEngine/FlaxEngine#1585]
**Issue description:**
I tried to serialize the ``class MySettings``. 

```csharp
public class MySettings : BaseSettings<BaseClass> { }

public class BaseSettings<T>
{
    public List<T> Data = new List<T>();
}

public class BaseClass { }

public class DrivenA : BaseClass { }

public class DrivenB : DrivenA { }
```

On deserialization, this error is thrown:

```
[Warning] Exception has been thrown during FlaxEngine.JsonAsset, 64a5ad30479d1d42ede1b690104b634a, D:\Projects/Flax/Test Blank/Content/My Settings.json. Type specified in JSON 'Game.DrivenB, Game.CSharp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' is not compatible with 'Game.BaseClass, Game.CSharp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. Path 'Data[0].$type', line 4, position 39.
```

BUT this is only thrown if flax had it freshly recompiled. If for example: You save the settings. Close Flax, reopen flax, you can open the json asset editor with no error. If you close the json asset editor, hit recompile, you will be prompted with the error again if you try to open the json asset.

In the class ``Newtonsoft.Json.Serialization.JsonSerializerInternalReader@826`` is the check that failed:
The check ``IsAssignableFrom`` fails and throws this error. If you inspect the assemblies of both types, even they have the same name, a comparison check will return false. It looks like the types have 2 different assemblies, even they should have the same assembly, maybe from recompiling? 

**Steps to reproduce:**

- Open attached project.
- Open Content/My Settings
- Click button to add elements, click save (Inspect Elements under Data)
- Close window
- Open Content/My Settings
- No Error, you can inspect the elements, everything is fine.
- Close window
- File -> Recompile Scripts
- Open Content/My Settings
- The Error should appear
- If you add now elements, save and close and open flax again, you can open the settings without error but only as long you don't recompile the scripts.

**Minimal reproduction project:**
[Test Blank.zip](https://github.com/FlaxEngine/FlaxEngine/files/12785198/Test.Blank.zip)

**Flax version:**
master 5fe135e

**Comment by Menotdan** [comment]
I had a similar bug actually, but it seems to go away for me permanently after re-launch. Recompiling scripts does not re-trigger my issue.

**Comment by schmidt-florian** [comment]
Some information about it:

Json.NET tries to get the assembly via `Assembly.LoadWithPartialName(assemblyName);` (DefaultSerializationBindercs:78)
However, this returns null. If I run the same code with the same name in a script, it returns the correct assembly.

I find this behavior very odd. As if there were 2 different app domains....

**Comment by SinnersSum** [comment]
I'm running into this same issue right now. I have a type `IDamageState` and a type `SimpleDamage` which implements it. When the serializer deserializes an asset containing this type however, it produces this exact error.

I presume it has something to do with domain reload as checking with a debugger reveals that `SimpleDamage` implements a `IDamageState` *but* it is from a different version of the module than the `IDamageState` that the serializer is checking for.

**Comment by SinnersSum** [comment]
I've begun working to address the serialization issue. My current strategy is to reset the serialization cache when the scripting assemblies are reloaded. We'll see if that's sufficient.

**Comment by schmidt-florian** [comment]
I was already there. I even changed the engine code and reset the cache. I don't remember all, but I think I had then trouble in other ways. I think I even replaced Json.NET with some other and had issues.

Maybe you have more luck than I had.

**Comment by SinnersSum** [comment]
Yeah unfortunately just clearing the cache hasn't fixed the issue. I'll try and figure out how it's resolving domains

**Comment by mafiesto4** [comment]
Thanks for reporting this. We will work on fixing this.

**Comment by SinnersSum** [comment]
Since this is stalling my work I'll try and get a head start on the problem, if I make any progress I'll PR it (here's hoping)

**Comment by schmidt-florian** [comment]
> Yeah unfortunately just clearing the cache hasn't fixed the issue. I'll try and figure out how it's resolving domains

Yeah, i already digged down this as well. But I am good in C# stuff. C++ NOPE and then the coupling of these two. I inspected it and decided... Flax Dev Team is way better at that than I am.



> Thanks for reporting this. We will work on fixing this.

Thank you, as always.


> Since this is stalling my work I'll try and get a head start on the problem, if I make any progress I'll PR it (here's hoping)

Yeah, I can see that this Issue can be deal-breaking. I found a workaround so I was satisfied with the classification "Future". Because Mafiesto has enough other stuff to do, I did not bother with this. But I can see that you need this Issue fixed. I mean it is only a matter of time before I will run into this issue again. Because I like to create complex abstract class scenarios ^^

What I have learned from this, they do it the right way the hot reloading, but somehow the old versions get stuck. I hope its "only" a small bug, but they probably need time to find it.

**Comment by SinnersSum** [comment]
> Yeah, i already digged down this as well. But I am good in C# stuff. C++ NOPE and then the coupling of these two. I inspected it and decided... Flax Dev Team is way better at that than I am.

Well the good news is the problem is probably entirely on the C# side; it was relatively easy to insert a callback into the domain-reload process, and Json.NET is a .NET serialization framework, so I don't think it's touching any of the native components except through their managed interfaces.

As to possible solutions, I'm looking at `ISerializationBinder`, which provides the mechanism by which Json.NET resolves type names into types, if anything is loading out of date domains it's probably that. I notice that flax is using the default one, so perhaps it's a matter of configuring it to make it domain-reload aware.

**Comment by SinnersSum** [comment]
I've patched around the problem, will link a PR when it's written.

**Comment by mafiesto4** [comment]
Fixed in #1888
